### ISSUE TITLE
Cannot open browser for login when running Instant app.

### ISSUE THREAD

**Issue by pvtai** [openid/AppAuth-Android#630]
I have enabled Instant app for my app, but I cannot open browser for input user/pass for login. After call performAuthorizationRequest() I don't receive anything, It working OK on Installed app but not Instant app.

**Comment by amrfarid140** [comment]
Can you provide a sample to reproduce the issue?

**Comment by pvtai** [comment]
> Can you provide a sample to reproduce the issue?

```
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.text.TextUtils;
import android.util.Log;

import androidx.annotation.AnyThread;
import androidx.annotation.ColorInt;
import androidx.annotation.MainThread;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.WorkerThread;
import androidx.browser.customtabs.CustomTabsIntent;

import net.openid.appauth.AppAuthConfiguration;
import net.openid.appauth.AuthState;
import net.openid.appauth.AuthorizationException;
import net.openid.appauth.AuthorizationRequest;
import net.openid.appauth.AuthorizationService;
import net.openid.appauth.AuthorizationServiceConfiguration;
import net.openid.appauth.ClientAuthentication;
import net.openid.appauth.ClientAuthentication.UnsupportedAuthenticationMethod;
import net.openid.appauth.EndSessionRequest;
import net.openid.appauth.ResponseTypeValues;
import net.openid.appauth.TokenResponse;
import net.openid.appauth.browser.BrowserMatcher;
import net.openid.appauth.browser.BrowserWhitelist;
import net.openid.appauth.browser.VersionedBrowserMatcher;
import net.openid.appauth.connectivity.ConnectionBuilder;

import java.io.IOException;
import java.lang.ref.WeakReference;
import java.net.HttpURLConnection;
import java.util.Objects;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.atomic.AtomicReference;

/**
 * The SDK's main interaction point with an application. Other operations will stem from this
 * class in order to provide a single interface for callers.
 */
@SuppressWarnings("WeakerAccess")
public class SsoAppAuth {

    private static final String TAG = "SsoAppAuth";

    private static final AtomicReference<WeakReference<SsoAppAuth>> INSTANCE_REF =
            new AtomicReference<>(new WeakReference<SsoAppAuth>(null));

    protected AtomicReference<AuthorizationService> mAuthService = new AtomicReference<>();
    protected final AuthStateManager mAuthStateManager;
    protected final OAuthClientConfiguration mConfiguration;

    protected final AtomicReference<SsoAuthListener> mInitializationListener = new AtomicReference<>();
    protected final AtomicReference<String> mClientId = new AtomicReference<>();
    protected final AtomicReference<AuthorizationRequest> mAuthRequest = new AtomicReference<>();
    protected final AtomicReference<CustomTabsIntent> mAuthIntent = new AtomicReference<>();

    protected ExecutorService mExecutor;

    @ColorInt
    protected int mCustomTabColor;

    protected ConnectionBuilder mConnectionBuilder;

    protected WeakReference<Context> mContext;

    /**
     * Retrieve the manager object via the static {@link WeakReference} or construct a new instance.
     *
     * @param context The Context from which to get the application's environment
     * @return am SsoAppAuth object
     */
    @AnyThread
    public static SsoAppAuth getInstance(@NonNull Context context) {
        Log.e(TAG, "getInstance");
        SsoAppAuth ssoAppAuth = INSTANCE_REF.get().get();
        if (ssoAppAuth == null) {
            ssoAppAuth = new SsoAppAuth(context.getApplicationContext());
            INSTANCE_REF.set(new WeakReference<>(ssoAppAuth));
        }

        return ssoAppAuth;
    }

    /**
     * Constructs an SsoAppAuth object. Provided the Context to initialize any other components.
     *
     * @param context The application Context
     */
    @AnyThread
    protected SsoAppAuth(Context context) {
        Log.e(TAG, "SsoAppAuth");
        mContext = new WeakReference<>(context);
        mExecutor = Executors.newSingleThreadExecutor();
        mAuthStateManager = AuthStateManager.getInstance(context.getApplicationContext());
        mConfiguration = OAuthClientConfiguration.getInstance(context);
    }

    /**
     * Initializes the SsoAppAuth object. This will fetch an OpenID Connect discovery document
     * from the issuer in the configuration to configure this instance for use. This method
     * will not customize the CustomTabs session. If you would like to customize the
     * CustomTabs session, use {@link #init(Context, SsoAuthListener, int)}
     *
     * @param context  The application context
     * @param listener An SsoAuthSuccessListener that will be called once the initialization is
     *                 complete
     */
    @AnyThread
    public void init(
            final Context context,
            final SsoAuthListener listener) {
        init(context, listener, 0);
    }

    /**
     * Initializes the OktaAppAuth object. This will fetch an OpenID Connect discovery document
     * from the issuer in the configuration to configure this instance for use.
     *
     * @param context        The application context
     * @param listener       An OktaAuthListener that will be called once the initialization is
     *                       complete
     * @param customTabColor The color that will be passed to
     *                       {@link CustomTabsIntent.Builder#setToolbarColor(int)}
     */
    @AnyThread
    public void init(
            final Context context,
            final SsoAuthListener listener,
            @ColorInt int customTabColor) {
        Log.e(TAG, "init");
        init(context, listener, customTabColor, DefaultSsoConnectionBuilder.INSTANCE);
    }

    /**
     * Initializes the SsoAppAuth object. This will fetch an OpenID Connect discovery document
     * from the issuer in the configuration to configure this instance for use.
     *
     * @param context        The application context
     * @param listener       An SsoAuthListener that will be called once the initialization is
     *                       complete
     * @param customTabColor The color that will be passed to
     *                       {@link CustomTabsIntent.Builder#setToolbarColor(int)}
     */
    @AnyThread
    public void init(
            final Context context,
            final SsoAuthListener listener,
            @ColorInt int customTabColor,
            final SsoConnectionBuilder ssoConnectionBuilder) {
        mCustomTabColor = customTabColor;
        mConnectionBuilder = new ConnectionBuilder() {
            @NonNull
            @Override
            public HttpURLConnection openConnection(@NonNull Uri uri) throws IOException {
                Log.d(TAG, "openConnection");
                return ssoConnectionBuilder.openConnection(uri);
            }
        };
        mExecutor.submit(new Runnable() {
            @Override
            public void run() {
                doInit(context, mConnectionBuilder, listener);
            }
        });
    }

    /**
     * Performs revocation of accessToken or refreshToken.
     *
     * @param token accessToken or refreshToken {@link SsoAppAuth#getTokens()}
     * @param listener revocation callback {@link RevokeTokenRequest.RevokeListener}
     */
    @SuppressWarnings("unused")
    public void revoke(final String token, @NonNull final RevokeTokenRequest.RevokeListener listener) {
        if (mConfiguration.hasConfigurationChanged()) {
            throw new IllegalStateException("Okta Configuration has changed");
        }
        if (mAuthStateManager.getCurrent().getAuthorizationServiceConfiguration() == null) {
            throw new IllegalStateException("Okta should be initialized first");
        }
        mExecutor.submit(new Runnable() {
            @Override
            public void run() {
                doRevoke(token, listener);
            }
        });
    }

    /**
     * Performs revocation of accessToken and refreshToken if they are available.
     *
     * @param listener revocation callback {@link RevokeTokenRequest.RevokeListener}
     */
    @SuppressWarnings("unused")
    public void revoke(@NonNull final RevokeTokenRequest.RevokeListener listener) {

        if (!isUserLoggedIn()) {
            throw new IllegalStateException("No logged in user found");
        }
        if (mConfiguration.hasConfigurationChanged()) {
            throw new IllegalStateException("Okta Configuration has changed");
        }
        if (mAuthStateManager.getCurrent().getAuthorizationServiceConfiguration() == null) {
            throw new IllegalStateException("Okta should be initialized first");
        }

        if (mAuthStateManager.getCurrent().getRefreshToken() != null) {
            //if we have refresh token we have to perform revoke it first
            mExecutor.submit(new Runnable() {
                @Override
                public void run() {
                    doRevoke(
                            mAuthStateManager.getCurrent().getRefreshToken(),
                            new RevokeTokenRequest.RevokeListener() {
                                @Override
                                public void onSuccess() {
                                    doRevoke(mAuthStateManager
                                                    .getCurrent().getAccessToken(),
                                            listener);
                                }

                                @Override
                                public void onError(AuthorizationException ex) {
                                    listener.onError(ex);
                                }
                            });
                }
            });
        } else {
            mExecutor.submit(new Runnable() {
                @Override
                public void run() {
                    doRevoke(mAuthStateManager.getCurrent().getAccessToken(), listener);
                }
            });
        }

    }

    @WorkerThread
    private void doRevoke(String token,@NonNull RevokeTokenRequest.RevokeListener listener) {
        RevokeTokenRequest request =
                new RevokeTokenRequest.Builder(
                        Objects.requireNonNull(Objects.requireNonNull(mAuthStateManager.getCurrent()
                                .getAuthorizationServiceConfiguration())
                                .discoveryDoc).docJson)
                        .addClientId(mClientId.get())
                        .addToken(token)
                        .addConnectionBuilder(mConnectionBuilder)
                        .build();

        if (request != null) {
            request.performRequest(listener);
        }
    }

    /**
     * Logs in a user and acquires authorization tokens for that user.
     *
     * @param context          The application context
     * @param completionIntent The PendingIntent to direct the flow upon successful completion
     * @param cancelIntent     The PendingIntent to direct the flow upon cancellation or failure
     */
    public void login(
            final Context context,
            final PendingIntent completionIntent,
            final PendingIntent cancelIntent) {
        login(context, completionIntent, cancelIntent, null);
    }

    /**
     * Logs in a user and acquires authorization tokens for that user. Uses a login hint provided
     * within method
     *
     * @param context          The application context
     * @param completionIntent The PendingIntent to direct the flow upon successful completion
     * @param cancelIntent     The PendingIntent to direct the flow upon cancellation or failure
     * @param payload          Additional request payload that should be sent during authorization
     */
    public void login(
            final Context context,
            final PendingIntent completionIntent,
            final PendingIntent cancelIntent,
            final AuthenticationPayload payload) {
        if (mConfiguration.hasConfigurationChanged()) {
            throw new IllegalStateException("Sso Configuration has changed");
        } else if (mAuthStateManager.getCurrent().getAuthorizationServiceConfiguration() == null) {
            throw new IllegalStateException("Sso should be initialized first");
        } else {
            mExecutor.submit(new Runnable() {
                @Override
                public void run() {
                    doLogin(
                            SsoManagementActivity.createStartIntent(
                                    context.getApplicationContext(),
                                    completionIntent,
                                    cancelIntent),
                            cancelIntent, payload);
                }
            });
        }
    }

    /**
     * Removes all stored information on current session like
     * Tokens and Authentication Server config.
     * NOTE: After removal {@link SsoAppAuth#init} should be called.
     */
    public void clearSession() {
        // discard the authorization and token state, but retain the configuration and
        // dynamic client registration (if applicable), to save from retrieving them again.
        AuthState currentState = mAuthStateManager.getCurrent();
        if (currentState.getAuthorizationServiceConfiguration() != null) {
            AuthState clearedState =
                    new AuthState(currentState.getAuthorizationServiceConfiguration());
            if (currentState.getLastRegistrationResponse() != null) {
                clearedState.update(currentState.getLastRegistrationResponse());
            }
            mAuthStateManager.replace(clearedState);
            Log.e(TAG, "clearSession");
        }
    }

    /**
     * Disposes state that will not normally be handled by garbage collection. This should be
     * called when this service is no longer required, including when any owning activity is
     * paused or destroyed (i.e. in {@link android.app.Activity #onDestroy()}).
     */
    public void dispose() {
        if (mAuthService.get() != null) {
            mAuthService.get().dispose();
            mAuthService.set(null);
        }
        mInitializationListener.set(null);
    }

    /**
     * Determines whether a user is currently authorized given the current (or updated)
     * configuration.
     *
     * @return {@code true} if a user is logged in and the configuration hasn't changed;
     * {@code false} otherwise
     */
    @AnyThread
    public boolean isUserLoggedIn() {
        return mAuthStateManager.getCurrent().isAuthorized() &&
                !mConfiguration.hasConfigurationChanged() &&
                mAuthStateManager.getCurrent().getAuthorizationServiceConfiguration() != null;
    }

    /**
     * Determines whether there is a refresh token in the application's storage.
     *
     * @return {@code true} if a refresh token is present; {@code false} otherwise
     */
    @SuppressWarnings("unused")
    public boolean hasNoRefreshToken() {
        return mAuthStateManager.getCurrent().getRefreshToken() == null;
    }

    /**
     * Determines whether there is an access token in the application's storage.
     *
     * @return {@code true} if an access token is present; {@code false} otherwise
     */
    @SuppressWarnings("unused")
    public boolean hasAccessToken() {
        return mAuthStateManager.getCurrent().getAccessToken() != null;
    }

    /**
     * The expiration time of the current access token (if available), as milliseconds from the
     * UNIX epoch (consistent with {@link System#currentTimeMillis()}).
     *
     * @return Milliseconds from the UNIX epoch at which point the access token will expire
     */
    public Long getAccessTokenExpirationTime() {
        return mAuthStateManager.getCurrent().getAccessTokenExpirationTime();
    }

    public String getScope() {
        return mAuthStateManager.getCurrent().getScope();
    }

    /**
     * Refreshes the access token if a refresh token is available to do so. This method will
     * do nothing if there is no refresh token.
     *
     * @param listener An SsoAuthSuccessListener that will be called once the refresh is complete
     */
    @SuppressWarnings("unused")
    public void refreshAccessToken(final SsoAuthListener listener) {
        if (hasNoRefreshToken()) {
            Log.d(TAG, "Calling refreshAccessToken without a refresh token");
            listener.onTokenFailure(AuthorizationException.TokenRequestErrors.INVALID_REQUEST);
            return;
        }

        ClientAuthentication clientAuthentication;
        try {
            clientAuthentication = mAuthStateManager.getCurrent().getClientAuthentication();
        } catch (UnsupportedAuthenticationMethod ex) {
            Log.e(TAG, "Token request cannot be made; client authentication for the token "
                    + "endpoint could not be constructed (%s)", ex);
            listener.onTokenFailure(AuthorizationException.TokenRequestErrors.INVALID_REQUEST);
            return;
        }
        if (mAuthStateManager.getCurrent().getLastAuthorizationResponse() == null) {
            Log.e(TAG, "No authorization configuration available for refresh request");
            listener.onTokenFailure(AuthorizationException.TokenRequestErrors.INVALID_REQUEST);
            return;
        }

        createAuthorizationServiceIfNeeded().performTokenRequest(
                mAuthStateManager.getCurrent().createTokenRefreshRequest(),
                clientAuthentication,
                new AuthorizationService.TokenResponseCallback() {
                    @Override
                    public void onTokenRequestCompleted(@Nullable TokenResponse tokenResponse,
                                                        @Nullable AuthorizationException
                                                                authException) {
                        handleAccessTokenResponse(
                                tokenResponse,
                                authException,
                                listener);
                    }
                });
    }

    /**
     * Determines whether there is an ID token in the application's storage.
     *
     * @return {@code true} if an ID token is present; {@code false} otherwise
     */
    public boolean hasIdToken() {
        return mAuthStateManager.getCurrent().getIdToken() != null;
    }

    /**
     * provides tokens for current logged in user.
     *
     * @return tokens {@link Tokens}
     */
    public Tokens getTokens() {
        return Tokens.fromAuthState(mAuthStateManager.getCurrent());
    }

    @MainThread
    private void doInit(final Context context, final ConnectionBuilder connectionBuilder, final SsoAuthListener listener) {
        mInitializationListener.set(listener);
        recreateAuthorizationService(context);

        if (mConfiguration.hasConfigurationChanged()) {
            // discard any existing authorization state due to the change of configuration
            Log.d(TAG, "Configuration change detected, discarding old state");
            mAuthStateManager.replace(new AuthState());
            if (!mConfiguration.isValid()) {
                Log.e(TAG, "Configuration was invalid: " + mConfiguration.getConfigurationError());
                listener.onTokenFailure(
                        AuthorizationException.GeneralErrors.INVALID_DISCOVERY_DOCUMENT);
            }

            mConfiguration.acceptConfiguration();
        }


        if (mAuthStateManager.getCurrent().getAuthorizationServiceConfiguration() != null) {
            // configuration is already created, skip to client initialization
            Log.d(TAG, "auth config already established");
            initializeClient();
            return;
        }

        Log.d(TAG, "Retrieving OpenID discovery doc");
        AuthorizationServiceConfiguration.fetchFromUrl(
                mConfiguration.getDiscoveryUri(),
                new AuthorizationServiceConfiguration.RetrieveConfigurationCallback() {
                    @Override
                    public void onFetchConfigurationCompleted(
                            @Nullable AuthorizationServiceConfiguration serviceConfiguration,
                            @Nullable AuthorizationException ex) {
                        Log.d(TAG, "onFetchConfigurationCompleted");
                        handleConfigurationRetrievalResult(serviceConfiguration, ex);
                    }
                },
                connectionBuilder);
    }

    /*
     * Initiates the client ID from the configuration.
     */
    @MainThread
    private void initializeClient() {
        Log.d(TAG, "Using static client ID: " + mConfiguration.getClientId());
        // use a statically configured client ID
        mClientId.set(mConfiguration.getClientId());
        initializeAuthRequest();
    }


    @MainThread
    private void initializeAuthRequest() {
        Log.d(TAG, "initializeAuthRequest");
        createAuthRequest(null);
        mInitializationListener.get().onSuccess();
    }

    private void createAuthRequest(@Nullable AuthenticationPayload payload) {
        Log.d(TAG, "Creating authorization request");
        AuthorizationRequest.Builder authRequestBuilder = new AuthorizationRequest.Builder(
                Objects.requireNonNull(mAuthStateManager.getCurrent().getAuthorizationServiceConfiguration()),
                mClientId.get(),
                ResponseTypeValues.CODE,
                mConfiguration.getRedirectUri())
                .setScopes(mConfiguration.getScopes());
        if (payload != null) {
            authRequestBuilder.setAdditionalParameters(payload.getAdditionalParameters());
            if (!TextUtils.isEmpty(payload.toString())) {
                authRequestBuilder.setState(payload.getState());
            }
            if (!TextUtils.isEmpty(payload.getLoginHint())) {
                authRequestBuilder.setLoginHint(payload.getLoginHint());
            }
        }
        mAuthRequest.set(authRequestBuilder.build());
    }

    private AuthorizationRequest createLoginRequest(@Nullable AuthenticationPayload payload) {
        Log.d(TAG, "Creating login request");
        AuthorizationRequest.Builder authRequestBuilder = new AuthorizationRequest.Builder(
                Objects.requireNonNull(mAuthStateManager.getCurrent().getAuthorizationServiceConfiguration()),
                mClientId.get(),
                ResponseTypeValues.CODE,
                mConfiguration.getRedirectUri())
                .setScopes(mConfiguration.getScopes())
                .setPrompt(mConfiguration.getPrompt());
        if (payload != null) {
            authRequestBuilder.setAdditionalParameters(payload.getAdditionalParameters());
            if (!TextUtils.isEmpty(payload.toString())) {
                authRequestBuilder.setState(payload.getState());
            }
            if (!TextUtils.isEmpty(payload.getLoginHint())) {
                authRequestBuilder.setLoginHint(payload.getLoginHint());
            }
        }
        return authRequestBuilder.build();
    }

    @MainThread
    private void warmUpBrowser(Uri uri) {
        Log.d(TAG, "Warming up browser instance for auth request");
        boolean isBrowserInstalled = PackageStoreManager.isBrowserInstalled(mContext.get());
        Log.d(TAG, "isBrowserInstalled: " + isBrowserInstalled);
        if (isBrowserInstalled) {
            CustomTabsIntent.Builder intentBuilder =
                    createAuthorizationServiceIfNeeded().createCustomTabsIntentBuilder(uri);
            intentBuilder.setToolbarColor(mCustomTabColor);
            CustomTabsIntent customTabsIntent = intentBuilder.build();
            customTabsIntent.intent.setFlags(Intent.FLAG_ACTIVITY_NO_HISTORY);
            customTabsIntent.intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
            mAuthIntent.set(customTabsIntent);
        } else {
            if (PolicyManager.isProfileOwnerApp(mContext.get())) {
                PackageStoreManager.openPlayStore(mContext.get(), "Firefox");
            } else {
                PackageStoreManager. openPlayStore(mContext.get(), "chrome");
            }
        }
    }

    @MainThread
    private void handleConfigurationRetrievalResult(AuthorizationServiceConfiguration config,
                                                    AuthorizationException ex) {
        if (config == null) {
            Log.e(TAG, "Failed to retrieve discovery document", ex);
            mInitializationListener.get().onTokenFailure(ex);
            return;
        }

        Log.d(TAG, "Discovery document retrieved");
        mAuthStateManager.replace(new AuthState(config));
        mExecutor.submit(new Runnable() {
            @Override
            public void run() {
                initializeClient();
            }
        });
    }

    @MainThread
    private void recreateAuthorizationService(Context context) {
        Log.d(TAG, "recreateAuthorizationService");
        if (mAuthService.get() != null) {
            Log.d(TAG, "Discarding existing AuthService instance");
            mAuthService.get().dispose();
        }
        mAuthService.set(createAuthorizationService(context));
        mAuthRequest.set(null);
        mAuthIntent.set(null);
    }

    /**
     * Handles recreating the authorization service if it has been cleared out.
     *
     * @return a usable instance of {@see AuthorizationService}
     */
    AuthorizationService createAuthorizationServiceIfNeeded() {
        if (mAuthService.get() == null) {
            Log.d(TAG, "createAuthorizationServiceIfNeeded mAuthService.get() == null");
            Context context = mContext.get();
            if (context == null) {
                throw new IllegalStateException("Invalid context. " +
                        "Okta should be initialized first");
            } else {
                recreateAuthorizationService(context);
            }
        }
        Log.d(TAG, "createAuthorizationServiceIfNeeded mAuthService.get() != null");
        return mAuthService.get();
    }

    private AuthorizationService createAuthorizationService(Context context) {
        Log.d(TAG, "Creating authorization service");
        @NonNull
        BrowserMatcher mBrowserMatcher;
        if (PolicyManager.isProfileOwnerApp(mContext.get())) {
            mBrowserMatcher = new BrowserWhitelist(
                    VersionedBrowserMatcher.FIREFOX_BROWSER,
                    VersionedBrowserMatcher.FIREFOX_CUSTOM_TAB
            );
        } else {
            mBrowserMatcher = new BrowserWhitelist(
                    VersionedBrowserMatcher.CHROME_BROWSER,
                    VersionedBrowserMatcher.CHROME_CUSTOM_TAB
            );
        }
        AppAuthConfiguration configuration = new AppAuthConfiguration.Builder()
                .setBrowserMatcher(mBrowserMatcher)
                .setConnectionBuilder(mConnectionBuilder)
                .build();
        return new AuthorizationService(context, configuration);
    }

    @MainThread
    private void doLogin(PendingIntent completionIntent,
                         PendingIntent cancelIntent,
                         AuthenticationPayload payload) {
        Log.d(TAG, "Starting authorization flow");
        AuthorizationRequest request = createLoginRequest(payload);
        warmUpBrowser(request.toUri());
        createAuthorizationServiceIfNeeded().performAuthorizationRequest(
                request,
                completionIntent,
                cancelIntent,
                mAuthIntent.get());
    }

    /**
     * Ends session of the current user within Okta.
     *  @param context          The application context
     * @param completionIntent The PendingIntent to direct the flow upon successful completion
     * @param cancelIntent     The PendingIntent to direct the flow upon cancellation or failure
     * @param idToken
     */
    public void signOutFromSso(
            final Context context,
            final PendingIntent completionIntent,
            final PendingIntent cancelIntent,
            final OnSigOutCallback onSigOutCallback,
            final String idToken) {
        if (!isUserLoggedIn()) {
            throw new IllegalStateException("No logged in user found");
        }
        mExecutor.submit(new Runnable() {
            @Override
            public void run() {
                doEndSession(
                        SsoManagementActivity.createStartIntent(
                                context,
                                completionIntent,
                                cancelIntent),
                        cancelIntent, onSigOutCallback, idToken);
            }
        });

    }

    @MainThread
    private void doEndSession(PendingIntent completionIntent, PendingIntent cancelIntent, OnSigOutCallback onSigOutCallback, String idToken) {
        Log.d(TAG, "Starting end session flow");

        EndSessionRequest request = new EndSessionRequest(
                Objects.requireNonNull(mAuthStateManager.getCurrent().getAuthorizationServiceConfiguration()),
                idToken,
                mConfiguration.getEndSessionRedirectUri());

        warmUpBrowser(request.toUri());

        CustomTabsIntent.Builder intentBuilder =
                createAuthorizationServiceIfNeeded()
                        .createCustomTabsIntentBuilder(request.toUri());
        intentBuilder.setToolbarColor(mCustomTabColor);
        CustomTabsIntent endSessionIntent = intentBuilder.build();

        createAuthorizationServiceIfNeeded()
                .performEndOfSessionRequest(request, completionIntent,
                        cancelIntent, endSessionIntent);

        onSigOutCallback.onSuccess();
    }

    @WorkerThread
    private void handleAccessTokenResponse(
            @Nullable TokenResponse tokenResponse,
            @Nullable AuthorizationException authException,
            @NonNull SsoAuthListener listener) {
        mAuthStateManager.updateAfterTokenResponse(tokenResponse, authException);
        if (authException == null) {
            listener.onSuccess();
        } else {
            Log.d(TAG, "Encountered an error with the access token response", authException);
            listener.onTokenFailure(authException);
        }
    }

    /**
     * Listener for OktaNativeAuth operations.
     */
    public interface OktaNativeAuthListener {
        /**
         * Called when the operation is successful to allow the caller to be notified.
         */
        void onSuccess();

        /**
         * Called when a failure occurs during the operation related to the authorization flow.
         *
         * @param ex The exception describing the failure
         */
        void onTokenFailure(@NonNull AuthenticationError ex);
    }

    /**
     * Listener for SsoAppAuth operations.
     */
    public interface SsoAuthListener {
        /**
         * Called when the operation is successful to allow the caller to be notified.
         */
        void onSuccess();

        /**
         * Called when a failure occurs during the operation related to the authorization flow.
         *
         * @param ex The exception describing the failure
         */
        void onTokenFailure(@NonNull AuthorizationException ex);
    }

}
```

On UI, I will call this method: doLogin(PendingIntent completionIntent,
                         PendingIntent cancelIntent,
                         AuthenticationPayload payload) in SsoAppAuth

After run this method: warmUpBrowser(request.toUri());
and run this: createAuthorizationServiceIfNeeded().performAuthorizationRequest(
                request,
                completionIntent,
                cancelIntent,
                mAuthIntent.get());
I don't receive anything else.

**Comment by amrfarid140** [comment]
Thanks @pvtai this is better but I meant a sample project with the instant app boilerplate setup. Would you be able to link to a git repo for your project for example?

**Comment by pvtai** [comment]
> Thanks @pvtai this is better but I meant a sample project with the instant app boilerplate setup. Would you be able to link to a git repo for your project for example?

Thank you for your reply, but my prj not public on repo.
I think you can check on this prj of you by "Enable Instant" app module, it'll cannot retrie browser list and cannot open any browser for input user/pass

**Comment by pvtai** [comment]
> Thanks @pvtai this is better but I meant a sample project with the instant app boilerplate setup. Would you be able to link to a git repo for your project for example?

I have created new repo at here https://github.com/pvtai/AppAuth
Step to test Instant app:
Step 1: Right click on "app" module and select "Refactor" -> "Enable Instant Apps Support..."
Step 2: Open Run/Debug Configurations and checked "Deploy as instant app" -> Apply -> OK
![InstantApp](https://user-images.githubusercontent.com/4888072/103056605-f5d73280-45cf-11eb-9d93-40c0c64d5729.png)

Step 3: Run app and check result.

**Comment by pvtai** [comment]
@amrfarid140 any update, please

**Comment by amrfarid140** [comment]
[The official docs state](https://developer.android.com/topic/google-play-instant/getting-started/instant-enabled-app-bundle#access-installed-apps) 

> When developing an instant experience, keep in mind that it cannot interact with installed apps on a device unless one of the following is true:
> - One or more activities within an installed app has set its android:visibleToInstantApps element to true – This element is available to apps running Android 8.0 (API level 26) or higher.
> - An installed app contains an intent filter that includes CATEGORY_BROWSABLE.
> - The instant experience is sending an intent using either the ACTION_SEND, ACTION_SENDTO, or ACTION_SEND_MULTIPLE action.

Which means that you have to provide your own BrowserDescriptor and CustomTabManager using this constructor for `AuthorizationService`

```
AuthorizationService(@NonNull Context context,
                         @NonNull AppAuthConfiguration clientConfiguration,
                         @Nullable BrowserDescriptor browser,
                         @NonNull CustomTabManager customTabManager)
```

**Comment by volo-droid** [comment]
> Which means that you have to provide your own BrowserDescriptor and CustomTabManager using this constructor for `AuthorizationService`

That constructor is not part of the public API (it's package-private)
